Python 3: file function -> open function

josephsl opened the pull request:
### Link to issue number:
Fixes #9038 

### Summary of the issue:
Python 3 retires file function in favor of using open function (with or without "with" statement).

### Description of how this pull request fixes the issue:
Edits to call "open" or "with open":

* NVDA helper and Watchdog: with file -> with open
* Add-on handler and update check: cPicle load/dump functions changed to use a "with open" block
* Config: use open/close for creating an empty config file
* Build version: use with open block
* Log viewer: a note about conversion eligibility and encoding issue notes.

### Testing performed:
Tested with source copy of NVDA (Python 3) along with source code version of latest master branch.

### Known issues with pull request:
As noted in log viewer comment, although log viewer/save log routine is eligible to be converted to "with open" block, there is a note about UTF-8 conversion, thus excluded for now.

### Change log entry:
None

josephsl committed:
Add-on handler and update check/Python 3: file function -> with open function for reading from and writing to pickle files. Re #9038.

josephsl committed:
NVDA helper and watchdog/Python 3: with file -> with open function. Re #9038.

josephsl committed:
Config/Python 3: use open function when creating an empty profile data file. Re #9038.

josephsl committed:
Build version/Python 3: use with open function when obtaining Git head/commit for version information construction. Re #9038.

Without modifying head/commit values for version info, NVDA will report 'year.major.minor' when running from source. Therefore edit this routine.

josephsl committed:
Log viewer/Python 3: add an explanatory note on with open function and encoding issue. Re #9038.

When saving logs, a log file object will be created. The routine responsible for this is eligible for conversion to 'with open' statement, but there is a note about UTF-8 codecs conversion problem. Thus add a note regarding this and exclude that routine from conversion for now.

michaelDCurran commented on source/gui/logViewer.py:
```diff
@@ -80,6 +80,7 @@ def onSaveAsCommand(self, evt):
 		try:
 			# codecs.open() forces binary mode, which is bad under Windows because line endings won't be converted to crlf automatically.
 			# Therefore, do the encoding manually.
+			# #9038 (Py3 review required): although eligible for "with open" function conversion, encoding issue must be settled first.
```
Per the documentation for open in Python3, providing no encoding causes it to default to locale.getpreferredencoding(False). On Windows (at least my machine) locale.getpreferredencoding(False) is cp1252 which only supports the Latin alphabet. I think we should provide encoding='utf-8' to the call to open, and then we can remove the .encode('utf-8') from the string when calling write.

michaelDCurran commented on source/addonHandler/__init__.py:
```diff
@@ -47,7 +47,8 @@ def loadState():
 	statePath=os.path.join(globalVars.appArgs.configPath,stateFilename)
 	try:
-		state = cPickle.load(file(statePath, "r"))
+		with open(statePath, "r") as f:
```
In Python 3 when pickling or unpickling objects, the file needs to be opened as binary so that no text encoding/decoding takes place. So for any open calls around pickle loads or dumps, the mode for reading must be rb and the mode for writing must be wb.

michaelDCurran reviewed (changes requested):
Based on the encoding changes for logViewer and pickling, I think this should go to threshold, not master. Unless of course @feerrenrut specifically asked for master for some reason?

josephsl commented:
Hi, @LeonardDer thinks it could be tested in master since the code is compatible between 2 and 3. I’ll wait until Reef asks using Threshold before proceeding. Thanks.

 

From: Michael Curran <notifications@github.com> 
Sent: Tuesday, June 4, 2019 4:51 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python 3: file function -> open function (#9663)

 

@michaelDCurran requested changes on this pull request.

Based on the encoding changes for logViewer and pickling, I think this should go to threshold, not master. Unless of course @feerrenrut <https://github.com/feerrenrut>  specifically asked for master for some reason?

  _____  

In source/gui/logViewer.py <https://github.com/nvaccess/nvda/pull/9663#discussion_r290535286> :

> @@ -80,6 +80,7 @@ def onSaveAsCommand(self, evt):
               try:
                       # codecs.open() forces binary mode, which is bad under Windows because line endings won't be converted to crlf automatically.
                       # Therefore, do the encoding manually.
+                      # #9038 (Py3 review required): although eligible for "with open" function conversion, encoding issue must be settled first.

Per the documentation for open in Python3, providing no encoding causes it to default to locale.getpreferredencoding(False). On Windows (at least my machine) locale.getpreferredencoding(False) is cp1252 which only supports the Latin alphabet. I think we should provide encoding='utf-8' to the call to open, and then we can remove the .encode('utf-8') from the string when calling write.

  _____  

In source/addonHandler/__init__.py <https://github.com/nvaccess/nvda/pull/9663#discussion_r290538225> :

> @@ -47,7 +47,8 @@ def loadState():
        global state
        statePath=os.path.join(globalVars.appArgs.configPath,stateFilename)
        try:
-              state = cPickle.load(file(statePath, "r"))
+              with open(statePath, "r") as f:

In Python 3 when pickling or unpickling objects, the file needs to be opened as binary so that no text encoding/decoding takes place. So for any open calls around pickle loads or dumps, the mode for reading must be rb and the mode for writing must be wb.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9663?email_source=notifications&email_token=AB4AXEAGPFAP4XRWFEJVYRDPY35XFA5CNFSM4HSDUUZ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2S56XI#pullrequestreview-245751645> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEBOQMRRPXXY6JDDRT3PY35XFANCNFSM4HSDUUZQ> .

josephsl commented:
Hi,

Actually, switched the base to Threshold and will take a look at log handler issue.

Thanks.

josephsl committed:
Add-on handler and update check/Python 3: use binary format when working with pickles. Re #9038.

Clarified by Mick Curran (NV Access): in Python 3, when loading and dumping pickles, binary format must be used (rb/wb).

josephsl committed:
Log viewer/Python 3: save log file in UTF-8 format. Re #9038.

Reviewed by Mick Curran (NV Access): open log file for saving with UTF-8 encoding from the start. Note that Python 2's open function does not include encoding parameter directly, but Python 3 does.

josephsl commented:
Hi,

All done.

Thanks.

leonardder commented:
I wasn't aware of the valid points Mick came up with. Feel free to ignore my request regardin this pr.

feerrenrut commented:
There are a few places we use ` codecs.open`, can you look into whether this still makes sense?

josephsl commented:
Hi, at least for now it may still work, although mick may have something for us to think about. Thanks.

 

From: Reef Turner <notifications@github.com> 
Sent: Thursday, June 6, 2019 9:44 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python 3: file function -> open function (#9663)

 

There are a few places we use codecs.open, can you look into whether this still makes sense?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9663?email_source=notifications&email_token=AB4AXEBNOV5U7W5FKBHEVFLPZE5EPA5CNFSM4HSDUUZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXDOS5A#issuecomment-499575156> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEBYP7FR5COUQ3DWIWDPZE5EPANCNFSM4HSDUUZQ> .
